### PR TITLE
[receiver/kafkareceiver] Add AVRO encoding support

### PR DESCRIPTION
**Pull request by thmshmm**
**Description:** Support to read AVRO encoded structured logs from Kafka

This features allows to read Kafka messages and decode them with any provided AVRO schema. It works the same way as the JSON unmarshaler and puts the decoded messaged into the log record body. From there it can be further processed with e.g. a transform processor.

**Link to tracking Issue:** #21067 

**Testing:** 
- Unit tests
- Manual testing via local Kafka, custom build OTel collector, producer application

**Documentation:** Added how to configure avro encoder on the kafka receiver.

### CODE REVIEW

File: receiver/kafkareceiver/config.go
Comment (atoulme) [inline receiver/kafkareceiver/config.go:42]: Is that schema a reference to a file path?

Comment (atoulme) [reply receiver/kafkareceiver/config.go:42]: It looks like it's a complex JSON object encoded as string, is that correct?

**Comment by atoulme** [comment]
Changes look good to me. Needs a review by codeowner.

File: receiver/kafkareceiver/config.go
Comment (thmshmm) [reply receiver/kafkareceiver/config.go:42]: yes, correct. Initially it was a file path but as I remember, @MovieStoreGuy prefered to have it inlined so I changed it.

**Comment by atoulme** [comment]
@MovieStoreGuy @pavolloffay please review as codeowners

**Comment by thmshmm** [comment]
@atoulme @MovieStoreGuy, I have added the extension as well. I think the extension only makes sense to unmarshal as we would not have a mapping from a otel log record back into any provided AVRO schema.

Please review and let me know if this matches your expectations.

**Comment by dmitryax** [comment]
@thmshmm can you please submit separate PRs for the kafka receiver and the new encoding extension?

**Comment by thmshmm** [comment]
@dmitryax see #31923. PRs splitted.

**Review by dmitryax — APPROVED** [review]

**Review by VihasMakwana — APPROVED** [review]

**Review by pavolloffay — APPROVED** [review]

**Comment by atoulme** [comment]
Now that https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/31923 has landed, maybe this PR can be simplified?

**Comment by thmshmm** [comment]
we could instantiate and execute the extension within the unmarshaler so that the unmarshaler just becomes a wrapper for now.

File: receiver/kafkareceiver/config.go
Comment (atoulme) [inline receiver/kafkareceiver/config.go:84]: imo the avro encoding should have the config defined on itself, and we should refer to the encoding extension ID. See https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/exporter/fileexporter/config.go#L45 for inspiration

Comment (thmshmm) [reply receiver/kafkareceiver/config.go:84]: is someone already working on supporting the encoding extensions in the kafka receiver or exporter? the changes introduced by this pr are more or less temporary and can be skipped if it could be used soon. but i would prefer not mix both intents.

Comment (atoulme) [reply receiver/kafkareceiver/config.go:84]: I was hoping to count on you, as you seem like you have a good handle on things.

Comment (thmshmm) [reply receiver/kafkareceiver/config.go:84]: sure, i'll have a look.

**Pull request closed without merging**
